### DISCUSSION TITLE
Event as std::variant

### DISCUSSION THREAD

**Discussion by TobiasWallner** [jupyter-xeus/cpp-terminal#276] [General]
How about we change the class [`Event`](https://github.com/jupyter-xeus/cpp-terminal/blob/master/cpp-terminal/event.hpp#L37-L42) into an `std::variant`?

The way I understand it is that only ever one of the member variables of the `Event` class will be used at the same time. So we might as well save some memory here by using a `variant`.

**Comment by flagarde** [comment]
Std::variant it's c++17 and we would like to be c++11. We can use a union and refactor the event class for this. I tried to do this before but I focused on the more low level stuffs before polishing this

**Reply by TobiasWallner** [reply]
ok

**Reply by TobiasWallner** [reply]
and can we have some kind of access operators with const ref and ref access or pointer access with either a `std::optional` or a pointer returned from a `get_if_<type>` function (does not need to be templated). I do not want to construct (malloc and copy) an `std::string` if i can get a reference to it. I would offer it in a pull request and you can see if you like it.

**Reply by flagarde** [reply]
Same problem this is c++17
